dcp-0007/dcp-0007.mediawiki

```diff
@@ -0,0 +1,193 @@
+<pre>
+DCP: 0007
+Title: Revert Treasury Expenditure Policy
+Author: Matheus Degiovani <contact@matheusd.com>
+Status: Active
+Created: 2021-06-23
+License: CC0-1.0
+License-Code: ISC
+Requires: DCP0006
+</pre>
+
+==Abstract==
+
+This proposes reverting the currently active maximum expenditure policy of the
+treasury account to the one specified in the original Politeia proposal for the
+decentralized treasury.
+
+==Motivation==
+
+The current algorithm that calculates the maximum expenditure allowed by a set
+of treasury spend transactions within a given window of blocks takes into
+account the average spent in a preceding window.
+
+This is different then the policy that was originally proposed and approved by
+stakeholders in the original Politeia proposal for development of the
+decentralized treasury<ref>[[https://proposals-archive.decred.org/proposals/c96290a]]</ref>
+and has caused an issue that has prevented payments from the treasury account
+from being performed.
+
+This proposal modifies the consensus rules to ensure the maximum expenditure
+policy is in line with what was explicitly approved via Politeia voting, by
+using the total income into the treasury as the reference value to determine the
+spending limit.
+
+The maximum expenditure amount is currently calculated by the <code>dcrd</code> reference
+node implementation by the
+<code>maxTreasuryExpenditure()</code><ref>[[https://github.com/decred/dcrd/blob/afff2fdbcd4c57ade4f0d13e78ad2d3efaebcdec/blockchain/treasury.go#L650]]</ref>
+function. This function defines the following policy:
+
+<blockquote>The sum of treasury spend amounts inside the most recent expenditure window
+cannot exceed the average of the spent amount in the preceding N windows in
+addition to a 50% increase. If no amount is spent in the preceding windows,
+the maximum allowed amount equals a bootstrap amount in addition to a 50%
+increase.</blockquote>
+
+For mainnet, the expenditure policy window has a size of 6912 blocks and 6
+preceding windows are checked, for a total of 41472 blocks. A bootstrap amount
+of 16K DCR is used.
+
+A deeper overview of the issue can be read on the 
+[[https://blog.decred.org/2021/06/25/Treasury-Expenditure-Policy-Bug/|Treasury Expenditure Policy Bug]]
+blog post.
+
+==Specification==
+
+Change the maximum expenditure policy to the following:
+
+<blockquote>The sum of treasury spend amounts inside the most recent expenditure window
+cannot exceed the total amount received by the treasury in the same window, in
+addition to a 50% increase.</blockquote>
```

Comment:
This needs to be more explicit, in my opinion.  Specifically, "a 50% increase" is an ambiguous amount.  e.g. Does the calculation use T-division, F-division, R-division, or C-division?  What are the exact floating point semantics (which largely differ between languages)?

All of that ambiguity can be avoided by explicitly providing the mathematical formula including the floor function to ensure integer arithmetic and positive values where the various division methods produce the same quotient.  Further, different implementations will then not be required to know the specific intricacies of the Go runtime.

Reply:
Github doesn't accept math in comments, so I've rendered and attached a complete formulation for the policy. Hopefully the definitions at the start are sufficient to define it unambiguously. I'll generate individual images for the terms/formulas once this is nailed down.

![dcp0007-math](https://user-images.githubusercontent.com/1752493/124005831-ed611980-d9af-11eb-8792-b6a3b5060e37.png)

Reply:
That looks much better.  The only part that is still slightly ambiguous to me is what happens when `h < pw`?  Is it clamped to 0, or considered illegal resulting in `S_{spent} = 0` and `S_{added} = 0`?  I realize this doesn't matter on mainnet, but it will on simnet.

I believe the correct behavior should be clamping it zero, in which case you probably want:

`S_{spent} = \sum_{i=max(h-pw, 0)}^{h-1} \beta_{i}^{ts}`

![image](https://user-images.githubusercontent.com/2115102/124039812-44271d00-d9c9-11eb-98eb-fd90965f75dc.png)

Similarly for `S_{added}`.

Also, I noticed `S_{spent}` in the equation and `S_{spend}` in the informal description.

Reply:
Updated.

![image](https://user-images.githubusercontent.com/1752493/124115681-7a55b280-da44-11eb-9628-b209752a24a9.png)

Reply:
With my crude math, on mainnet, it looks like this allows for spending per window, not counting treasury adds:

12,948dcr now
10,933dcr next year
5,502dcr in five years
2,361dcr in ten years
430dcr in twenty years

Look right?

Reply:
Those are about right, yeah.

Reply:
Updated to use svg renderings of the equations.

Reply:
One thing that is somewhat obvious, but might be worth calling out, is whether the "sum of the transactions" refers to the sum of the inputs or outputs in each case (e.g., in the case of treasury add and treasury base, the sum refers to the transaction outputs being added to the treasury, but in the case of treasury spends the sum refers to the transaction inputs being spent from the treasury which includes fees, [as accounted for in your draft reference implementation](https://github.com/decred/dcrd/blob/1b04ee6c1f4569d53ba398ea2e6e20478dadc71f/blockchain/treasury.go#L631)).

Reply:
Tweaked the relevant section to the below. Do you think this is sufficient?

![tx_types](https://user-images.githubusercontent.com/1752493/126184060-e3b50e65-11fb-44d0-9517-e27c56346e13.png)

Reply:
Yeah, that looks good to me.